File: Shortest_Path/baekjoon_11780.cpp
#include<iostream>
#include<vector>
#define inf 1e9

int graph[101][101];
int path[101][101] ={0, };

int main(){
    int n, m;
    std::cin >> n >> m;


    for(int i = 1; i<=n; i ++){
        for(int j = 1; j<=n; j++){
            if(i == j){
                graph[i][j] = 0;
            }
            else{graph[i][j] = inf;}
        }
    }

    for(int i = 1; i <= m; i++){
        int a,b,c;
        std::cin >> a >> b >> c;
        if(graph[a][b] > c){
            graph[a][b] = c;
            path[a][b] = a;
        }
    }

    for(int Mid = 1; Mid <= n; Mid++){
        for(int Start = 1; Start <= n; Start++){
            if(Start == Mid) continue;
            for(int End = 1; End <= n; End++){
                if(End == Mid) continue;
                if(graph[Start][End] > graph[Start][Mid] + graph[Mid][End]){
                    graph[Start][End] = graph[Start][Mid] + graph[Mid][End];
                    path[Start][End] = path[Mid][End];
                }
            }
        }
    }

    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= n; j++){
            if(graph[i][j] == inf){
                std::cout << 0 << " ";
            }
            else{
                std::cout << graph[i][j] << " ";
            }
        }
        std::cout << "\n";
    }

    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= n; j++){
            if(graph[i][j] == inf || i==j){
                std::cout << 0 <<std::endl;
            }
            else{
                std::vector<int> ans;
                ans.push_back(j);
                int cur = j;
                while(cur != i){
                    ans.push_back(path[i][cur]);
                    cur = path[i][cur];
                }
                std::cout << ans.size() << " ";
                for(int k = ans.size() - 1; k >= 0; k--){
                    std::cout<<ans[k] << " ";
                }
                std::cout<<"\n";
            }
        }
    }

    return 0;

}

File: Shortest_Path/baekjoon_11780.py
import sys
read = sys.stdin.readline
inf = sys.maxsize
        
def solution(n,m,graoh,path):
    
    for mid in range(1,n+1):
        for start in range(1,n+1):
            if start == mid:
                continue
            for end in range(1, n+1):
                if end == mid:
                    continue
                
                if graph[start][end]> graph[start][mid] + graph[mid][end]:
                    graph[start][end] = graph[start][mid] + graph[mid][end]
                    path[start][end] = path[mid][end]
                    
    for i in range(1,n+1):
        for j in range(1,n+1):
            print(0 if graph[i][j] == inf else graph[i][j], end = ' ')
        print()
    
    for start in range(1,n+1):
        for end in range(1,n+1):
            if graph[start][end] == inf or start == end:
                print(0)
                continue
            else:
                ans=[end]
                while end!=start:
                    ans.append(path[start][end])
                    end = path[start][end]
                    
                print(len(ans), end = " ")
                for p in ans[::-1]:
                    print(p, end = " ")
                print()
                
    
if __name__ == "__main__":
    n = int(input())
    m = int(input())
    graph = [[inf]*(n+1) for _ in range(n+1)]
    path = [[0]*(n+1) for _ in range(n+1)]
    
    for i in range(1,n+1):
        graph[i][i] = 0
    
    for _ in range(m):
        a,b,c = map(int, input().split())
        if c < graph[a][b]:
            graph[a][b] = c
            path[a][b] = a
            
    solution(n,m,graph,path)
        
